llvm/include/llvm/Support/TypeSize.h

```diff
@@ -97,17 +97,29 @@ template <typename LeafTy, typename ValueTy> class FixedOrScalableQuantity {
 
   constexpr FixedOrScalableQuantity() = default;
   constexpr FixedOrScalableQuantity(ScalarTy Quantity, bool Scalable)
-      : Quantity(Quantity), Scalable(Scalable) {}
+      : Quantity(Quantity), Scalable(Quantity ? Scalable : false) {}
```

Comment:
As discussed, this might be a little dangerous right now because it'll cause functions like `isScalable` to return false when given `TypeSize::getScalable(0)`.  There's also other operators like `*` which can produce zero without going via this constructor.

It might be better to ensure all the interfaces return canonical results for fixed and scalable zero rather than changing the way zero is stored.

Whilst canonicalisation is the correct way forward if you don't specifically need it at this time then it's best removed and done in isolation.

Reply:
Thanks, I've removed the canonicalisation now.